scylla/src/client/session.rs

```diff
@@ -1253,41 +1251,91 @@ impl Session {
         &self,
         statement: &Statement,
     ) -> Result<PreparedStatement, PrepareError> {
-        let cluster_state = self.get_cluster_state();
-        let connections_iter = cluster_state.iter_working_connections_to_nodes()?;
-
-        // Prepare statements on all connections concurrently
-        let handles = connections_iter.map(|c| async move { c.prepare_raw(statement).await });
-        let mut results = join_all(handles).await.into_iter();
-
-        // If at least one prepare was successful, `prepare()` returns Ok.
-        // Find the first result that is Ok, or Err if all failed.
-
-        // Safety: there is at least one node in the cluster, and `Cluster::iter_working_connections_to_nodes()`
-        // returns either an error or an iterator with at least one connection, so there will be at least one result.
-        let first_ok: Result<RawPreparedStatement, RequestAttemptError> =
-            results.by_ref().find_or_first(Result::is_ok).unwrap();
-        let mut prepared: PreparedStatement = first_ok
-            .map_err(|first_attempt| PrepareError::AllAttemptsFailed { first_attempt })?
-            .into_prepared_statement();
-
-        // Validate prepared ids equality
-        for statement in results.flatten() {
-            if prepared.get_id() != &statement.prepared_response.id {
-                return Err(PrepareError::PreparedStatementIdsMismatch);
-            }
+        /// Prepares the statement on either all nodes or all shards.
+        ///
+        /// Returns:
+        /// - `Ok(RawPreparedStatement))`, if preparation succeeded on at least one connection;
+        /// - `Err(PrepareError)`, if no connection is working or preparation failed on all attempted connections.
+        async fn prepare_on_all(
+            statement: &Statement,
+            cluster_state: &ClusterState,
+            on_all_shards: bool,
+        ) -> Result<PreparedStatement, PrepareError> {
+            // Find the first result that is Ok, or Err if all failed.
+            let working_connections = if on_all_shards {
+                itertools::Either::Left(cluster_state.iter_working_connections_to_shards()?)
+            } else {
+                itertools::Either::Right(cluster_state.iter_working_connections_to_nodes()?)
+            };
+
+            let preparations =
+                working_connections.map(|c| async move { c.prepare_raw(statement).await });
+            let raw_prepared_statements_results = join_all(preparations).await;
+
+            let mut raw_prepared_statements_results_iter =
+                raw_prepared_statements_results.into_iter();
+
+            // Safety: We pass a nonempty iterator, so there will be at least one result.
+            let first_ok_or_error = raw_prepared_statements_results_iter
+                .by_ref()
+                .find_or_last(|res| res.is_ok())
+                .unwrap(); // Safety: there is at least one connection.
```

Comment:
Commit: "session: prepare: introduce fallback to all shards "

Previously `find_or_first` was used, now you changed it to `find_or_last`. Why?